src/java.base/share/classes/java/io/Console.java

```diff
@@ -659,20 +659,24 @@ private static Console instantiateConsole(boolean istty) {
              * providers are available, or instantiation failed, java.base built-in
              * Console implementation is used.
              */
-            PrivilegedAction<Console> pa = () -> {
-                var consModName = System.getProperty("jdk.console",
-                        JdkConsoleProvider.DEFAULT_PROVIDER_MODULE_NAME);
-                return ServiceLoader.load(ModuleLayer.boot(), JdkConsoleProvider.class).stream()
-                        .map(ServiceLoader.Provider::get)
-                        .filter(jcp -> consModName.equals(jcp.getClass().getModule().getName()))
-                        .map(jcp -> jcp.console(istty, CHARSET))
-                        .filter(Objects::nonNull)
-                        .findAny()
-                        .map(jc -> (Console) new ProxyingConsole(jc))
-                        .orElse(null);
-            };
-            c = AccessController.doPrivileged(pa);
-        } catch (ServiceConfigurationError ignore) {
+            c = AccessController.doPrivileged(new PrivilegedAction<Console>() {
+                public Console run() {
+                    var consModName = System.getProperty("jdk.console",
+                            JdkConsoleProvider.DEFAULT_PROVIDER_MODULE_NAME);
+
+                    for (var jcp : ServiceLoader.load(ModuleLayer.boot(), JdkConsoleProvider.class)) {
+                        if (consModName.equals(jcp.getClass().getModule().getName())) {
+                            var jc = jcp.console(istty, CHARSET);
+                            if (jc != null) {
+                                return new ProxyingConsole(jc);
+                            }
+                            break;
```

Comment:
```suggestion
```
The original `findAny` is only after `filter(Objects::nonNull)` meaning we don't return `null` on a null `jcp.console` result.

Reply:
Yes, `break` guarantees that the search completes one way or another once the module name has been matched. This is not how it used to be done.

Reply:
Right. Since `findAny` is after the module name matching, there is at most 1 match. In the case we didn't find any, the final `orElse(null)` eventually returns null. So the refactored code is semantically the same.

Reply:
It's only semantically the same if we assume a module can only provide a single `JdkConsoleProvider`, no? The `break;` disallows multiple providers (for disjoint sets of charsets) in a single module.

Reply:
Claes has described the issue well. Like I said, `break` short-circuits the search. If a module can provide more than one console provider, the suggested stream-less replacement is **not** equivalent.

If a module can provide more than one console provider, not only the code needs to be fixed, but a relevant test should be added too. The test should be in this PR, but tagged with the initial bug, [8295803], not this (performance) bug.

[8295803]: https://bugs.openjdk.org/browse/JDK-8295803

Reply:
In fact, this started simply for incorporating JLine implementation into Console, and using ServiceLoader was a mere means to load its impl as it resides outside the java.base module. So yes, module and its console implementation is 1:1, which is reflected by the system property `jdk.console` that takes the module name. So that `break;` effectively shortcut the unnecessary looping (I don't think it would happen btw).
That said, I think it needs to be described in the comment above that piece of code. I will add it shortly.

Reply:
So, it's the previous (stream) version that was more permissive and inadvertently so, yes?

Reply:
Yes, correct. If hypothetically Jline provided two impls, it were not specified which impl was used. Now we only use the first one found.